### ISSUE TITLE
JSON Input: Constraints

### ISSUE THREAD

**Issue by dgasmith** [leeping/geomeTRIC#26]
It would be good if the JSON constraint input could be a list of tuples rather than a string. This would make the interface a bit more canonical JSON and lower the barrier to create this input.

**Comment by dgasmith** [comment]
Can you use:

```
if isinstance(constraint, str):
    ...
elif isinstance(constraint, (tuple, list)):
    ...
else:
   raise TypeError(...)
```

**Comment by leeping** [comment]
I think that could work - @yudongqiu , any thoughts?

Thanks,

- Lee-Ping

**Comment by leeping** [comment]
Currently, an example of the constraints input string looks like this:

```
$freeze
xyz 1-5

$set
angle 2 1 5 110.0

$scan
dihedral 2 1 5 6 100.0 120.0 3 # A comment
dihedral 1 5 6 7 110.0 150.0 3 # A comment
```

This corresponds to a dictionary that looks like this:

```
{ 'freeze' : [('xyz', '1-5')], 'set' : [('angle', '2', '1', '5', '110.0')], 'scan' : [('dihedral', '2', '1', '5', '6', '100.0', '120.0', '3'), ('dihedral', '1', '5', '6', '7', '110.0', '150.0', '3')] }
```

I think we can easily take as input a dictionary like the one above.  It's much easier if we require all of the elements in each tuple to be "words" (i.e. strings without spaces).  If so, all we need to do is write an extra function to convert the dictionary to the constraints input string and pass that to `optimize.py::ParseConstraints()` without modifying the latter.

**Comment by leeping** [comment]
I understand the alternative is to rewrite `optimize.py::ParseConstraints()` so that it supports multiple input types, but I don't see how that is possible without substantially restructuring it.  That function is quite complicated because it supports simultaneous freezing / setting / scanning of various combinations of constraints.

**Comment by yudongqiu** [comment]
@leeping Thanks a lot for the clear specs of the new constraints input format. I will implement that and make a PR soon.

**Comment by leeping** [comment]
Addressed in PR #27 .
